src/lib/PageTracker.js

```diff
@@ -0,0 +1,347 @@
+    userId = '';
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * [constructor]
+     *
+     * @return {PageTracker} PageTracker instance
+     */
+    constructor() {
+        this.activeUserHandler = this.activeUserHandler.bind(this);
+        this.inactiveUserHandler = this.inactiveUserHandler.bind(this);
+        this.throttledActiveUserHandler = this.activeUserHandler().bind(this);
+    }
+
+    /**
+     * [destructor]
+     *
+     * @return {void}
+     */
+    destroy() {
+        this.stopTracking();
+        this.unbindIdleUserEvents();
+        this.clearIdleTimeout();
+        this.currentPage = null;
+        this.file = {};
+        this.sessionId = null;
+        this.startTime = null;
+    }
+
+    /**
+     * Get the session id
+     *
+     * @public
+     * @return {string}
+     */
+    getSessionId() {
+        return this.sessionId;
+    }
+
+    /**
+     * Call the interval starting function and set the idle timer
+     *
+     * @public
+     * @param {Object} contentInsightsOptions - Content Insights Related Options
+     * @param {Object} requestOptions - Contains host information and headers to make the request
+     * @param {Object} file - File that is being previewed
+     * @return {void}
+     */
+    init(contentInsightsOptions, requestOptions, file) {
+        // Generate a new session id. This would need to be later changed to be generated on server side
+        this.sessionId = uuidv4();
+
+        // Set User and file information
+        this.userEid = contentInsightsOptions.userEid ? contentInsightsOptions.userEid : '';
+        this.fileOwnerEid = contentInsightsOptions.ownerEID ? contentInsightsOptions.ownerEID : '';
+        this.userId = contentInsightsOptions.userID ? contentInsightsOptions.userID : '';
+
+        // Set hosting and file information
+        this.requestOptions = requestOptions;
+        this.file = file;
+
+        // Start tracking time on file if content insights is turn on for that particular file
+        if (contentInsightsOptions.contentInsightsConfig) {
+            this.isActive = contentInsightsOptions.contentInsightsConfig.isActive;
+            if (this.isActive) {
+                this.setIdleTimer();
+                this.startTracking();
+            }
+        }
+    }
+
+    /**
+     * Wheter Advanced Content Insights is on/off
+     *
+     * @public
+     * @return {boolean}
+     */
+    isContentInsightsActive() {
+        return this.isActive;
+    }
+
+    /**
+     * Start the content insights interval, to report data each number of seconds defined on a Const
+     *
+     * @public
+     * @return {void}
+     */
+    startTracking() {
+        this.stopTracking();
+        this.startTime = Date.now();
+
+        this.reportingIntervalId = setInterval(() => {
+            this.reportContentInsight();
```

Comment:
So if I'm reading this correctly, if the user stays on a page and is active (does something every 30s) this page tracker will continue to emit a report saying the user is on the page for 5s, on the page for 5s, on the page for 5s, etc...?

Reply:
Yes. ACI tracking is an optional feature (will be turned off for a file as default), so we won’t have a lot of files sending these requests. In our previous experience, we found that this approach was the best way to report these insights, because using browser-events to detect page unload or using the beacon API weren’t always reliable. Also when we looked how people accomplish this type of reporting, we found that they were doing the same back then.